notes/overriding.md

```diff
@@ -0,0 +1,675 @@
+---
+feature: overriding
+start-date: 2021-03-18
+author: Yann Hamdaoui
+---
+
+# Overriding
+
+This document is a proposal for an overriding mechanism in Nickel. It is
+expected to evolve while we put these ideas in practice, but shall serve as
+a design and implementation baseline.
+
+Related issues: [#103](https://github.com/tweag/nickel/issues/103),
+[#240](https://github.com/tweag/nickel/issues/240),
+[#255](https://github.com/tweag/nickel/issues/255),
+[#279](https://github.com/tweag/nickel/issues/279).
+
+## Context
+
+### Definition of the problem
+
+*Overriding* is the ability to take an existing configuration - in our case, a
+record - and update a specific part of it, such as a field or the field of a
+nested record.
+
+A naive and unergonomic of way of doing it is to repeat all the original
+values in a new record:
+
+```
+let record = {
+  a = 1,
+  b = "str",
+  c = false
+} in
+{
+  a = record.a,
+  b = record.b,
+  c = true
+}
+```
+
+This quickly become unmanageable for large records, which are common in Nix and
+Nixpkgs. In functional languages, this is usually better done using what's
+called functional update. Functional update is an operation that takes the
+original record, one or several field to update, the corresponding new value,
+and return a new updated record. It has the same semantics as our first snippet,
+but doesn't require to rewrite all unchanged fields.
+
+It can have a builtin syntax, such as OCaml's `with`: `{record with field =
+new_value}`, Haskell's `record {field = newValue}`, Nix `//` operator `record //
+{field = newValue}`, or Rust's syntax `RecordDataType {field: new_value,
+..record}`. There are more advanced programming techniques that make updating
+deeply nested records ergonomic such as
+[Lenses](https://www.fpcomplete.com/haskell/tutorial/lens/) in Haskell, but
+these rely too heavily on advanced language and typing features to be practical.
```

Comment:
It's also incorrect, in that we want some kind of “late binding” behaviour in recursive records. Maybe it should be explained.

Reply:
This is explained in the following section.